File: sdks/java/src/main/java/com/blok/blok/server/BlokNodeRuntimeService.java
package com.blok.blok.server;

import com.blok.blok.errors.BlokError;
import com.blok.blok.errors.BlokErrorCategory;
import com.blok.blok.errors.BlokErrorSeverity;
import com.blok.blok.node.NodeRegistry;
import com.blok.blok.types.Context;
import com.blok.blok.types.ExecutionRequest;
import com.blok.blok.types.ExecutionResult;
import com.blok.blok.types.NodeConfig;
import com.blok.blok.types.Request;
import com.blok.blok.types.Response;
// `java_multiple_files = true` in runtime.proto puts each message and enum
// at the top level of the `com.blok.runtime.v1` package.
import com.blok.runtime.v1.ErrorCategory;
import com.blok.runtime.v1.ErrorSeverity;
import com.blok.runtime.v1.ExecuteEvent;
import com.blok.runtime.v1.ExecuteRequest;
import com.blok.runtime.v1.ExecuteResponse;
import com.blok.runtime.v1.HealthRequest;
import com.blok.runtime.v1.HealthResponse;
import com.blok.runtime.v1.ListNodesRequest;
import com.blok.runtime.v1.ListNodesResponse;
import com.blok.runtime.v1.Metrics;
import com.blok.runtime.v1.NodeDescriptor;
import com.blok.runtime.v1.NodeError;
import com.blok.runtime.v1.NodeRuntimeGrpc;
import com.blok.runtime.v1.NodeStarted;
import com.blok.runtime.v1.RuntimeState;
import com.blok.runtime.v1.TriggerInfo;
import com.blok.runtime.v1.WorkflowInfo;
import com.google.protobuf.Timestamp;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.google.gson.JsonSyntaxException;
import com.google.gson.reflect.TypeToken;
import com.google.protobuf.ByteString;
import io.grpc.Status;
import io.grpc.stub.StreamObserver;

import java.lang.reflect.Type;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;

/**
 * gRPC implementation of the canonical Blok {@code NodeRuntime} v1 service.
 *
 * <p>Wire contract: {@code proto/blok/runtime/v1/runtime.proto}. Generated
 * stubs are produced by {@code protobuf-maven-plugin} at build time
 * (configured in {@code pom.xml}).
 *
 * <p>Single Responsibility: translate proto messages into the SDK's internal
 * {@link ExecutionRequest} / {@link ExecutionResult} and dispatch to
 * {@link NodeRegistry}. All node-level error handling lives in
 * {@code NodeRegistry.execute(ExecutionRequest)}.
 *
 * <p>The proto sends {@code inputs}, {@code previous_output}, {@code vars},
 * and the request {@code body} as raw JSON-encoded {@code bytes}. The SDK
 * JSON-decodes them lazily here at the proto boundary so node handlers see
 * the same map shapes as the HTTP path.
 */
public final class BlokNodeRuntimeService extends NodeRuntimeGrpc.NodeRuntimeImplBase {

    private static final Gson GSON = new GsonBuilder().serializeNulls().create();
    private static final Type MAP_TYPE = new TypeToken<Map<String, Object>>() {}.getType();

    private final NodeRegistry registry;
    private final String sdkVersion;

    public BlokNodeRuntimeService(NodeRegistry registry, String sdkVersion) {
        if (registry == null) {
            throw new IllegalArgumentException("registry must not be null");
        }
        this.registry = registry;
        this.sdkVersion = sdkVersion == null || sdkVersion.isBlank() ? "1.0.0" : sdkVersion;
    }

    @Override
    public void execute(ExecuteRequest request, StreamObserver<ExecuteResponse> responseObserver) {
        ExecutionRequest executionRequest;
        try {
            executionRequest = decodeExecuteRequest(request);
        } catch (DecodeException e) {
            responseObserver.onError(Status.INVALID_ARGUMENT.withDescription(e.getMessage()).asRuntimeException());
            return;
        }

        ExecutionResult result = registry.execute(executionRequest);
        responseObserver.onNext(encodeExecuteResponse(result, executionRequest.getNode().getName()));
        responseObserver.onCompleted();
    }

    /**
     * Server-streaming variant of {@link #execute}. Emits, in order:
     * (1) one {@code NodeStarted} event marking call acceptance,
     * (2) one terminal {@code ExecuteResponse} matching the unary payload.
     *
     * <p>Log capture (LogLine events) is intentionally out of scope for the
     * Phase 5 Java pilot — {@code NodeHandler.execute} has no per-call
     * logger sink, and threading one through would change the SDK API.
     * Real-time log streaming arrives in a follow-up.
     */
    @Override
    public void executeStream(ExecuteRequest request, StreamObserver<ExecuteEvent> responseObserver) {
        ExecutionRequest executionRequest;
        try {
            executionRequest = decodeExecuteRequest(request);
        } catch (DecodeException e) {
            responseObserver.onError(Status.INVALID_ARGUMENT.withDescription(e.getMessage()).asRuntimeException());
            return;
        }

        responseObserver.onNext(ExecuteEvent.newBuilder()
                .setStarted(NodeStarted.newBuilder().setAt(currentTimestamp()).build())
                .build());

        ExecutionResult result = registry.execute(executionRequest);
        ExecuteResponse finalResponse = encodeExecuteResponse(result, executionRequest.getNode().getName());

        responseObserver.onNext(ExecuteEvent.newBuilder().setFinal(finalResponse).build());
        responseObserver.onCompleted();
    }

    private static Timestamp currentTimestamp() {
        long millis = System.currentTimeMillis();
        return Timestamp.newBuilder()
                .setSeconds(millis / 1000)
                .setNanos((int) ((millis % 1000) * 1_000_000))
                .build();
    }

    @Override
    public void health(HealthRequest request, StreamObserver<HealthResponse> responseObserver) {
        HealthResponse.Builder builder = HealthResponse.newBuilder()
                .setStatus(HealthResponse.Status.SERVING)
                .setSdkVersion(sdkVersion);
        builder.addAllRegisteredNodes(registry.nodeNames());
        responseObserver.onNext(builder.build());
        responseObserver.onCompleted();
    }

    @Override
    public void listNodes(ListNodesRequest request, StreamObserver<ListNodesResponse> responseObserver) {
        ListNodesResponse.Builder builder = ListNodesResponse.newBuilder()
                .setSdkName("blok-java")
                .setSdkVersion(sdkVersion)
                .setProtoVersion("1.0.0");
        for (String name : registry.nodeNames()) {
            builder.addNodes(NodeDescriptor.newBuilder().setName(name).build());
        }
        responseObserver.onNext(builder.build());
        responseObserver.onCompleted();
    }

    // =========================================================================
    // Codec — proto ↔ internal types
    // =========================================================================

    private static final class DecodeException extends Exception {
        DecodeException(String message) { super(message); }
    }

    private static ExecutionRequest decodeExecuteRequest(ExecuteRequest req) throws DecodeException {
        if (!req.hasNode() || req.getNode().getName().isEmpty()) {
            throw new DecodeException("ExecuteRequest.node is required");
        }

        Map<String, Object> inputs = decodeJsonObject(req.getInputs(), "inputs");

        RuntimeState state = req.getState();
        Object previousOutput = decodeJsonValue(state.getPreviousOutput(), "previous_output");
        Map<String, Object> vars = decodeJsonObject(state.getVars(), "vars");

        TriggerInfo trigger = req.getTrigger();
        Object body = decodeRequestBody(trigger.getBody(), trigger.getHeadersMap());

        WorkflowInfo workflow = req.getWorkflow();

        Context ctx = new Context();
        ctx.setId(workflow.getRunId());
        ctx.setWorkflowName(workflow.getName());
        ctx.setWorkflowPath(workflow.getPath());

        Request internalRequest = new Request();
        internalRequest.setBody(body);
        internalRequest.setHeaders(new HashMap<>(trigger.getHeadersMap()));
        internalRequest.setParams(new HashMap<>(trigger.getParamsMap()));
        internalRequest.setQuery(new HashMap<>(trigger.getQueryMap()));
        internalRequest.setCookies(new HashMap<>(trigger.getCookiesMap()));
        internalRequest.setMethod(trigger.getMethod());
        internalRequest.setUrl(trigger.getUrl());
        internalRequest.setBaseUrl(trigger.getBaseUrl());
        ctx.setRequest(internalRequest);

        Response internalResponse = new Response();
        internalResponse.setData(previousOutput);
        internalResponse.setContentType("application/json");
        internalResponse.setSuccess(true);
        ctx.setResponse(internalResponse);

        ctx.setVars(vars);
        ctx.setEnv(new HashMap<>(state.getEnvMap()));

        NodeConfig nodeConfig = new NodeConfig();
        nodeConfig.setName(req.getNode().getName());
        nodeConfig.setType(req.getNode().getType());
        nodeConfig.setConfig(inputs);

        ExecutionRequest exec = new ExecutionRequest();
        exec.setNode(nodeConfig);
        exec.setContext(ctx);
        return exec;
    }

    private ExecuteResponse encodeExecuteResponse(ExecutionResult result, String nodeName) {
        ExecuteResponse.Builder builder = ExecuteResponse.newBuilder()
                .setSuccess(result.isSuccess())
                .setContentType("application/json");

        if (result.isSuccess() && result.getData() != null) {
            builder.setData(ByteString.copyFromUtf8(GSON.toJson(result.getData())));
        }

        if (result.getVars() != null && !result.getVars().isEmpty()) {
            builder.setVarsDelta(ByteString.copyFromUtf8(GSON.toJson(result.getVars())));
        }

        if (result.getMetrics() != null) {
            Metrics.Builder metrics = Metrics.newBuilder();
            if (result.getMetrics().getDurationMs() != null) {
                metrics.setDurationMs(result.getMetrics().getDurationMs());
            }
            if (result.getMetrics().getCpuMs() != null) {
                metrics.setCpuMs(result.getMetrics().getCpuMs());
            }
            if (result.getMetrics().getMemoryBytes() != null) {
                metrics.setMemoryBytes(result.getMetrics().getMemoryBytes());
            }
            builder.setMetrics(metrics.build());
        }

        if (!result.isSuccess()) {
            builder.setError(internalErrorToProto(result.getErrors(), nodeName));
        }

        return builder.build();
    }

    /**
     * Build a proto {@link NodeError} from whatever {@link ExecutionResult}
     * carried.
     *
     * <p>Two paths, both producing the same proto shape:
     * <ul>
     *   <li><b>Structured (preferred)</b> — {@code errVal} is a typed
     *       {@link BlokError}. All 19 fields serialize losslessly via
     *       {@link #blokErrorToProto}. Auto-fills
     *       {@code node}/{@code sdk}/{@code sdk_version}/{@code runtime_kind}
     *       if the BlokError didn't set them itself.</li>
     *   <li><b>Loose</b> — {@code errVal} is anything else (Map, String,
     *       null, Throwable). Wrapped via {@link BlokError#fromUnknown}
     *       (always produces {@code category=INTERNAL} with the original
     *       payload preserved in {@code details_json}) and then serialized
     *       via the structured path.</li>
     * </ul>
     */
    private NodeError internalErrorToProto(Object errVal, String nodeName) {
        BlokError.Origin origin = BlokError.Origin.defaults(nodeName, sdkVersion);
        if (errVal instanceof BlokError be) {
            be.applyOriginIfMissing(origin);
            return blokErrorToProto(be);
        }
        return blokErrorToProto(BlokError.fromUnknown(errVal, origin));
    }

    /**
     * Serialize a fully-populated {@link BlokError} into the proto wire
     * format. The cause chain is serialized as a list of proto NodeError
     * messages; each element's own {@code causes} list is left empty (the
     * chain is already flat at the BlokError layer).
     */
    private NodeError blokErrorToProto(BlokError err) {
        NodeError.Builder b = NodeError.newBuilder()
                .setCode(err.getCode() != null ? err.getCode() : "")
                .setCategory(categoryToProto(err.getCategory()))
                .setSeverity(severityToProto(err.getSeverity()))
                .setNode(err.getNode())
                .setSdk(err.getSdk())
                .setSdkVersion(err.getSdkVersion())
                .setRuntimeKind(err.getRuntimeKind())
                .setMessage(err.getMessage() != null ? err.getMessage() : "")
                .setDescription(err.getDescription())
                .setRemediation(err.getRemediation())
                .setDocUrl(err.getDocUrl())
                .setStack(err.getStack())
                .setHttpStatus(err.getHttpStatus())
                .setRetryable(err.isRetryable())
                .setRetryAfterMs(err.getRetryAfterMs());

        long epochSec = err.getAt().getEpochSecond();
        int nanos = err.getAt().getNano();
        b.setAt(Timestamp.newBuilder().setSeconds(epochSec).setNanos(nanos).build());

        if (err.getDetails() != null) {
            b.setDetailsJson(ByteString.copyFromUtf8(GSON.toJson(err.getDetails())));
        }
        if (err.getContextSnapshot() != null) {
            b.setContextSnapshotJson(ByteString.copyFromUtf8(GSON.toJson(err.getContextSnapshot())));
        }

        for (Map<String, Object> cause : err.getCauses()) {
            b.addCauses(causeMapToProto(cause));
        }
        return b.build();
    }

    /**
     * Convert one cause-chain link (already a snake_case map) into a proto
     * NodeError. Each link's own {@code causes} list is left empty; the chain
     * is already flat at the BlokError layer.
     */
    private NodeError causeMapToProto(Map<String, Object> cause) {
        BlokErrorCategory category = BlokErrorCategory.parse(stringField(cause, "category"));
        BlokErrorSeverity severity = BlokErrorSeverity.parse(stringField(cause, "severity"));
        NodeError.Builder b = NodeError.newBuilder()
                .setCode(stringField(cause, "code"))
                .setCategory(categoryToProto(category))
                .setSeverity(severityToProto(severity))
                .setNode(stringField(cause, "node"))
                .setSdk(stringField(cause, "sdk"))
                .setSdkVersion(stringField(cause, "sdk_version"))
                .setRuntimeKind(stringField(cause, "runtime_kind"))
                .setMessage(stringField(cause, "message"))
                .setDescription(stringField(cause, "description"))
                .setRemediation(stringField(cause, "remediation"))
                .setDocUrl(stringField(cause, "doc_url"))
                .setStack(stringField(cause, "stack"))
                .setHttpStatus(intField(cause, "http_status", 500))
                .setRetryable(boolField(cause, "retryable", false))
                .setRetryAfterMs(longField(cause, "retry_after_ms", 0L));

        Object atRaw = cause.get("at");
        if (atRaw instanceof String at) {
            try {
                java.time.Instant instant = java.time.Instant.parse(at);
                b.setAt(Timestamp.newBuilder()
                        .setSeconds(instant.getEpochSecond())
                        .setNanos(instant.getNano())
                        .build());
            } catch (Exception ignored) {
                // best effort
            }
        }

        Object det = cause.get("details");
        if (det != null) {
            b.setDetailsJson(ByteString.copyFromUtf8(GSON.toJson(det)));
        }
        Object snap = cause.get("context_snapshot");
        if (snap != null) {
            b.setContextSnapshotJson(ByteString.copyFromUtf8(GSON.toJson(snap)));
        }
        return b.build();
    }

    private static ErrorCategory categoryToProto(BlokErrorCategory c) {
        return switch (c) {
            case VALIDATION -> ErrorCategory.VALIDATION;
            case CONFIGURATION -> ErrorCategory.CONFIGURATION;
            case DEPENDENCY -> ErrorCategory.DEPENDENCY;
            case TIMEOUT -> ErrorCategory.TIMEOUT;
            case PERMISSION -> ErrorCategory.PERMISSION;
            case RATE_LIMIT -> ErrorCategory.RATE_LIMIT;
            case NOT_FOUND -> ErrorCategory.NOT_FOUND;
            case CONFLICT -> ErrorCategory.CONFLICT;
            case CANCELLED -> ErrorCategory.CANCELLED;
            case PROTOCOL -> ErrorCategory.PROTOCOL;
            case DATA -> ErrorCategory.DATA;
            case INTERNAL -> ErrorCategory.INTERNAL;
        };
    }

    private static ErrorSeverity severityToProto(BlokErrorSeverity s) {
        return switch (s) {
            case INFO -> ErrorSeverity.INFO;
            case WARN -> ErrorSeverity.WARN;
            case FATAL -> ErrorSeverity.FATAL;
            case ERROR -> ErrorSeverity.ERROR;
        };
    }

    private static String stringField(Map<String, Object> m, String key) {
        Object v = m.get(key);
        return v instanceof String s ? s : "";
    }

    private static int intField(Map<String, Object> m, String key, int fallback) {
        Object v = m.get(key);
        if (v instanceof Number n) return n.intValue();
        return fallback;
    }

    private static long longField(Map<String, Object> m, String key, long fallback) {
        Object v = m.get(key);
        if (v instanceof Number n) return n.longValue();
        return fallback;
    }

    private static boolean boolField(Map<String, Object> m, String key, boolean fallback) {
        Object v = m.get(key);
        if (v instanceof Boolean b) return b;
        return fallback;
    }

    /** Decode JSON-encoded bytes into a typed map. Empty bytes → empty map. */
    private static Map<String, Object> decodeJsonObject(ByteString bytes, String field) throws DecodeException {
        if (bytes.isEmpty()) return new HashMap<>();
        String json = bytes.toStringUtf8();
        try {
            JsonElement element = JsonParser.parseString(json);
            if (element.isJsonObject()) {
                Map<String, Object> map = GSON.fromJson(element, MAP_TYPE);
                return map != null ? map : new HashMap<>();
            }
            // Wrap non-object payloads under a reserved key so handlers
            // expecting a map don't crash.
            Map<String, Object> wrapped = new HashMap<>();
            wrapped.put("_value", GSON.fromJson(element, Object.class));
            return wrapped;
        } catch (JsonSyntaxException ex) {
            throw new DecodeException("invalid `" + field + "` JSON: " + ex.getMessage());
        }
    }

    /** Decode JSON-encoded bytes into an arbitrary value. Empty bytes → null. */
    private static Object decodeJsonValue(ByteString bytes, String field) throws DecodeException {
        if (bytes.isEmpty()) return null;
        try {
            return GSON.fromJson(bytes.toStringUtf8(), Object.class);
        } catch (JsonSyntaxException ex) {
            throw new DecodeException("invalid `" + field + "` JSON: " + ex.getMessage());
        }
    }

    /**
     * Decode the trigger body. JSON content-types parse as JSON; everything
     * else surfaces as a raw string for the node to interpret.
     */
    private static Object decodeRequestBody(ByteString bytes, Map<String, String> headers) {
        if (bytes.isEmpty()) return null;
        String contentType = pickHeader(headers, "content-type");
        if (contentType.toLowerCase().contains("application/json")) {
            try {
                return GSON.fromJson(bytes.toStringUtf8(), Object.class);
            } catch (JsonSyntaxException ignored) {
                // fall through
            }
        }
        return new String(bytes.toByteArray(), StandardCharsets.UTF_8);
    }

    private static String pickHeader(Map<String, String> headers, String name) {
        if (headers == null) return "";
        for (Map.Entry<String, String> entry : headers.entrySet()) {
            if (entry.getKey().equalsIgnoreCase(name)) return entry.getValue();
        }
        return "";
    }
}

File: sdks/php/src/Server/BlokNodeRuntimeService.php
<?php

declare(strict_types=1);

namespace Blok\Blok\Server;

use Blok\Blok\Errors\BlokError;
use Blok\Blok\Errors\BlokErrorCategory as InternalCategory;
use Blok\Blok\Errors\BlokErrorSeverity as InternalSeverity;
use Blok\Blok\Errors\Origin;
use Blok\Blok\Node\NodeRegistry;
use Blok\Blok\Types\Context as BlokContext;
use Blok\Blok\Types\ExecutionRequest as BlokExecutionRequest;
use Blok\Blok\Types\ExecutionResult;
use Blok\Blok\Types\NodeConfig;
use Blok\Blok\Types\Request as BlokRequest;
use Blok\Blok\Types\Response as BlokResponse;
use Blok\Runtime\V1\ErrorCategory;
use Blok\Runtime\V1\ErrorSeverity;
use Blok\Runtime\V1\ExecuteRequest;
use Blok\Runtime\V1\ExecuteResponse;
use Blok\Runtime\V1\HealthRequest;
use Blok\Runtime\V1\HealthResponse;
use Blok\Runtime\V1\HealthResponse\Status as HealthStatus;
use Blok\Runtime\V1\ListNodesRequest;
use Blok\Runtime\V1\ListNodesResponse;
use Blok\Runtime\V1\Metrics;
use Blok\Runtime\V1\NodeDescriptor;
use Blok\Runtime\V1\NodeError;
use Blok\Runtime\V1\RuntimeState;
use Blok\Runtime\V1\TriggerInfo;
use Blok\Runtime\V1\WorkflowInfo;
use Spiral\RoadRunner\GRPC\ContextInterface;
use Spiral\RoadRunner\GRPC\Exception\GRPCException;
use Spiral\RoadRunner\GRPC\Exception\InvokeException;
use Spiral\RoadRunner\GRPC\StatusCode;

/**
 * gRPC implementation of the canonical `blok.runtime.v1.NodeRuntime` v1 service.
 *
 * Owns a reference to a shared {@see NodeRegistry} so a single registry serves
 * both transports. Codec helpers sit at the proto boundary, leaving
 * {@see NodeRegistry::execute()} unchanged regardless of which transport
 * delivered the request.
 *
 * The proto sends `inputs`, `previous_output`, `vars`, and the request `body`
 * as raw JSON-encoded bytes; this class JSON-decodes them lazily.
 */
final class BlokNodeRuntimeService implements NodeRuntimeInterface
{
    private const SDK_NAME = 'blok-php';
    private const RUNTIME_KIND = 'runtime.php';
    private const PROTO_VERSION = '1.0.0';

    public function __construct(
        private readonly NodeRegistry $registry,
        private readonly string $sdkVersion = '1.0.0',
    ) {}

    public function Execute(ContextInterface $ctx, ExecuteRequest $in): ExecuteResponse
    {
        try {
            $internal = $this->decodeExecuteRequest($in);
        } catch (DecodeException $e) {
            throw new InvokeException($e->getMessage(), StatusCode::INVALID_ARGUMENT);
        }

        $nodeName = $internal->node->name;
        $result = $this->registry->execute($internal);

        return $this->encodeExecuteResponse($result, $nodeName);
    }

    public function Health(ContextInterface $ctx, HealthRequest $in): HealthResponse
    {
        return (new HealthResponse())
            ->setStatus(HealthStatus::SERVING)
            ->setSdkVersion($this->sdkVersion)
            ->setRegisteredNodes($this->registry->nodeNames());
    }

    public function ListNodes(ContextInterface $ctx, ListNodesRequest $in): ListNodesResponse
    {
        $descriptors = [];
        foreach ($this->registry->nodeNames() as $name) {
            $descriptors[] = (new NodeDescriptor())->setName($name);
        }

        return (new ListNodesResponse())
            ->setNodes($descriptors)
            ->setSdkName(self::SDK_NAME)
            ->setSdkVersion($this->sdkVersion)
            ->setProtoVersion(self::PROTO_VERSION);
    }

    // ===== Codec — proto <-> internal types =====

    /**
     * @throws DecodeException
     */
    public function decodeExecuteRequest(ExecuteRequest $req): BlokExecutionRequest
    {
        $node = $req->getNode();
        if ($node === null || $node->getName() === '') {
            throw new DecodeException('ExecuteRequest.node is required');
        }

        $inputs = self::decodeJsonObject($req->getInputs(), 'inputs');
        $state = $req->getState() ?? new RuntimeState();
        $trigger = $req->getTrigger() ?? new TriggerInfo();
        $workflow = $req->getWorkflow() ?? new WorkflowInfo();

        $previousOutput = self::decodeJsonValue($state->getPreviousOutput(), 'previous_output');
        $vars = self::decodeJsonObject($state->getVars(), 'vars');
        $headers = self::mapToArray($trigger->getHeaders());
        $body = self::decodeRequestBody($trigger->getBody(), $headers);

        $request = new BlokRequest(
            body: $body,
            headers: $headers,
            params: self::mapToArray($trigger->getParams()),
            query: self::mapToArray($trigger->getQuery()),
            method: $trigger->getMethod(),
            url: $trigger->getUrl(),
            cookies: self::mapToArray($trigger->getCookies()),
            baseUrl: $trigger->getBaseUrl(),
        );

        $response = new BlokResponse(
            data: $previousOutput,
            contentType: 'application/json',
            success: true,
            error: null,
        );

        $context = new BlokContext(
            id: $workflow->getRunId(),
            workflowName: $workflow->getName(),
            workflowPath: $workflow->getPath(),
            request: $request,
            response: $response,
            vars: $vars,
            env: self::mapToArray($state->getEnv()),
        );

        $nodeConfig = new NodeConfig(
            name: $node->getName(),
            type: $node->getType(),
            config: $inputs,
        );

        return new BlokExecutionRequest(node: $nodeConfig, context: $context);
    }

    public function encodeExecuteResponse(ExecutionResult $result, string $nodeName): ExecuteResponse
    {
        $response = (new ExecuteResponse())
            ->setSuccess($result->success)
            ->setContentType('application/json');

        if ($result->success && $result->data !== null) {
            $response->setData(self::encodeJsonBytes($result->data));
        }

        if (is_array($result->vars) && $result->vars !== []) {
            $response->setVarsDelta(self::encodeJsonBytes($result->vars));
        }

        if ($result->metrics !== null) {
            $response->setMetrics(
                (new Metrics())
                    ->setDurationMs((float) $result->metrics->durationMs)
                    ->setMemoryBytes((int) $result->metrics->memoryBytes),
            );
        }

        if (! $result->success) {
            $response->setError($this->internalErrorToProto($result->errors, $nodeName));
        }

        return $response;
    }

    /**
     * Build a proto {@see NodeError} from whatever {@see ExecutionResult}
     * carried.
     *
     * Two paths, both producing the same proto shape:
     * - **Structured (preferred)** — `$err` is a typed {@see BlokError}.
     *   All 19 fields serialize losslessly via {@see self::blokErrorToProto()}.
     *   Auto-fills `node`/`sdk`/`sdk_version`/`runtime_kind` if the BlokError
     *   didn't set them itself.
     * - **Loose** — `$err` is anything else (array, string, null, Throwable).
     *   Wrapped via {@see BlokError::fromUnknown()} (always produces
     *   `category=INTERNAL` with the original payload preserved in
     *   `details_json`) and then serialized via the structured path.
     */
    public function internalErrorToProto(mixed $err, string $nodeName): NodeError
    {
        $origin = Origin::defaults($nodeName, $this->sdkVersion);
        if ($err instanceof BlokError) {
            $err->applyOriginIfMissing($origin);
            return $this->blokErrorToProto($err);
        }
        return $this->blokErrorToProto(BlokError::fromUnknown($err, $origin));
    }

    /**
     * Serialize a fully-populated {@see BlokError} into the proto wire
     * format. The cause chain is serialized as a list of proto NodeError
     * messages; each element's own `causes` list is left empty (the chain
     * is already flat at the BlokError layer).
     */
    private function blokErrorToProto(BlokError $err): NodeError
    {
        $node = (new NodeError())
            ->setCode($err->errorCode)
            ->setCategory(self::categoryToProto($err->category))
            ->setSeverity(self::severityToProto($err->severity))
            ->setNode($err->node)
            ->setSdk($err->sdk)
            ->setSdkVersion($err->sdkVersion)
            ->setRuntimeKind($err->runtimeKind)
            ->setMessage($err->getMessage())
            ->setDescription($err->description)
            ->setRemediation($err->remediation)
            ->setDocUrl($err->docUrl)
            ->setStack($err->stack)
            ->setHttpStatus($err->httpStatus)
            ->setRetryable($err->retryable)
            ->setRetryAfterMs($err->retryAfterMs);

        $node->setAt(self::timeToProto($err->at));
        if ($err->details !== null) {
            $node->setDetailsJson(self::encodeJsonBytes($err->details));
        }
        if ($err->contextSnapshot !== null) {
            $node->setContextSnapshotJson(self::encodeJsonBytes($err->contextSnapshot));
        }

        $causes = [];
        foreach ($err->causes as $causePayload) {
            $causes[] = self::causeArrayToProto($causePayload);
        }
        $node->setCauses($causes);
        return $node;
    }

    /**
     * Convert one cause-chain link (already a snake_case array) into a proto
     * {@see NodeError}. Each link's own `causes` list is left empty.
     *
     * @param array<string, mixed> $cause
     */
    private static function causeArrayToProto(array $cause): NodeError
    {
        $category = InternalCategory::parse(self::strField($cause, 'category'));
        $severity = InternalSeverity::parse(self::strField($cause, 'severity'));
        $node = (new NodeError())
            ->setCode(self::strField($cause, 'code'))
            ->setCategory(self::categoryToProto($category))
            ->setSeverity(self::severityToProto($severity))
            ->setNode(self::strField($cause, 'node'))
            ->setSdk(self::strField($cause, 'sdk'))
            ->setSdkVersion(self::strField($cause, 'sdk_version'))
            ->setRuntimeKind(self::strField($cause, 'runtime_kind'))
            ->setMessage(self::strField($cause, 'message'))
            ->setDescription(self::strField($cause, 'description'))
            ->setRemediation(self::strField($cause, 'remediation'))
            ->setDocUrl(self::strField($cause, 'doc_url'))
            ->setStack(self::strField($cause, 'stack'))
            ->setHttpStatus(self::intField($cause, 'http_status', 500))
            ->setRetryable(self::boolField($cause, 'retryable', false))
            ->setRetryAfterMs(self::intField($cause, 'retry_after_ms', 0));

        $atRaw = $cause['at'] ?? null;
        if (is_string($atRaw) && $atRaw !== '') {
            try {
                $node->setAt(self::timeToProto(new \DateTimeImmutable($atRaw)));
            } catch (\Exception) {
                // best effort
            }
        }

        if (isset($cause['details']) && $cause['details'] !== null) {
            $node->setDetailsJson(self::encodeJsonBytes($cause['details']));
        }
        if (isset($cause['context_snapshot']) && $cause['context_snapshot'] !== null) {
            $node->setContextSnapshotJson(self::encodeJsonBytes($cause['context_snapshot']));
        }
        return $node;
    }

    private static function categoryToProto(InternalCategory $c): int
    {
        return match ($c) {
            InternalCategory::Validation => ErrorCategory::VALIDATION,
            InternalCategory::Configuration => ErrorCategory::CONFIGURATION,
            InternalCategory::Dependency => ErrorCategory::DEPENDENCY,
            InternalCategory::Timeout => ErrorCategory::TIMEOUT,
            InternalCategory::Permission => ErrorCategory::PERMISSION,
            InternalCategory::RateLimit => ErrorCategory::RATE_LIMIT,
            InternalCategory::NotFound => ErrorCategory::NOT_FOUND,
            InternalCategory::Conflict => ErrorCategory::CONFLICT,
            InternalCategory::Cancelled => ErrorCategory::CANCELLED,
            InternalCategory::Protocol => ErrorCategory::PROTOCOL,
            InternalCategory::Data => ErrorCategory::DATA,
            InternalCategory::Internal => ErrorCategory::INTERNAL,
        };
    }

    private static function severityToProto(InternalSeverity $s): int
    {
        return match ($s) {
            InternalSeverity::Info => ErrorSeverity::INFO,
            InternalSeverity::Warn => ErrorSeverity::WARN,
            InternalSeverity::Fatal => ErrorSeverity::FATAL,
            default => ErrorSeverity::ERROR,
        };
    }

    private static function timeToProto(\DateTimeImmutable $time): \Google\Protobuf\Timestamp
    {
        $ts = new \Google\Protobuf\Timestamp();
        $ts->setSeconds($time->getTimestamp());
        $ts->setNanos((int) ((int) $time->format('u') * 1000));
        return $ts;
    }

    /**
     * @param array<string, mixed> $m
     */
    private static function strField(array $m, string $key): string
    {
        $v = $m[$key] ?? null;
        return is_string($v) ? $v : '';
    }

    /**
     * @param array<string, mixed> $m
     */
    private static function intField(array $m, string $key, int $fallback): int
    {
        $v = $m[$key] ?? null;
        if (is_int($v)) return $v;
        if (is_float($v)) return (int) $v;
        if (is_string($v) && is_numeric($v)) return (int) $v;
        return $fallback;
    }

    /**
     * @param array<string, mixed> $m
     */
    private static function boolField(array $m, string $key, bool $fallback): bool
    {
        $v = $m[$key] ?? null;
        return is_bool($v) ? $v : $fallback;
    }

    /**
     * Decode JSON-encoded bytes into an associative array. Non-object payloads
     * are wrapped under `_value` so handlers expecting an array don't crash.
     *
     * The wire format always encodes maps as JSON objects (`{}`); this method
     * inspects the first non-whitespace byte to distinguish object from array,
     * since PHP cannot tell empty `{}` from empty `[]` after decoding.
     *
     * @return array<string, mixed>
     * @throws DecodeException
     */
    public static function decodeJsonObject(string $bytes, string $field): array
    {
        if ($bytes === '') {
            return [];
        }

        $parsed = json_decode($bytes, true);
        if (json_last_error() !== JSON_ERROR_NONE) {
            throw new DecodeException(sprintf('invalid `%s` JSON: %s', $field, json_last_error_msg()));
        }

        $firstChar = ltrim($bytes)[0] ?? '';
        if ($firstChar === '{' && is_array($parsed)) {
            return $parsed;
        }

        return ['_value' => $parsed];
    }

    /**
     * Decode JSON-encoded bytes into any value (object, array, scalar, null).
     *
     * @throws DecodeException
     */
    public static function decodeJsonValue(string $bytes, string $field): mixed
    {
        if ($bytes === '') {
            return null;
        }

        $parsed = json_decode($bytes, true);
        if (json_last_error() !== JSON_ERROR_NONE) {
            throw new DecodeException(sprintf('invalid `%s` JSON: %s', $field, json_last_error_msg()));
        }

        return $parsed;
    }

    /**
     * Decode the trigger body bytes. Parses as JSON when Content-Type is JSON;
     * otherwise returns the raw UTF-8 string.
     *
     * @param array<string, string> $headers
     */
    public static function decodeRequestBody(string $bytes, array $headers): mixed
    {
        if ($bytes === '') {
            return null;
        }

        $contentType = strtolower(self::pickHeader($headers, 'content-type'));
        if (str_contains($contentType, 'application/json')) {
            $parsed = json_decode($bytes, true);
            if (json_last_error() === JSON_ERROR_NONE) {
                return $parsed;
            }
            // fall through and return as raw string
        }

        return $bytes;
    }

    /**
     * Encode any value as a JSON byte string.
     */
    public static function encodeJsonBytes(mixed $value): string
    {
        $json = json_encode($value, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE);
        return $json === false ? '{}' : $json;
    }

    /**
     * Case-insensitive header lookup.
     *
     * @param array<string, string> $headers
     */
    public static function pickHeader(array $headers, string $name): string
    {
        $needle = strtolower($name);
        foreach ($headers as $key => $value) {
            if (strtolower((string) $key) === $needle) {
                return (string) $value;
            }
        }
        return '';
    }

    /**
     * Convert a proto MapField into a plain associative array.
     *
     * @return array<string, string>
     */
    private static function mapToArray(mixed $map): array
    {
        if ($map === null) {
            return [];
        }

        $out = [];
        foreach ($map as $k => $v) {
            $out[(string) $k] = (string) $v;
        }
        return $out;
    }
}
